### ISSUE TITLE
Use different concurrent queue implementation

### ISSUE THREAD

**Issue by erniejunior** [ZeroCM/zcm#259]
There is a queue implementation without locks which is provably performing well and it is thoroughly tested. I think I would prefer it over the current implementation.

https://github.com/cameron314/concurrentqueue

**Comment by jbendes** [comment]
I'm all game for the use of a better queue. I am not familiar enough with the concurrent queue to comment on its implementation. Aside from code quality and functionality, my top concern would be licensing and compatibility. The queue would have to work on any OS (I don't want to move farther from OS uniformity). It would have to not add any build dependencies (a key design principle of zcm is that it has _very_ limited required external dependencies). And last but not least it must have a license that will allow us to make it part of the repo which is covered under LGPL.

**Comment by erniejunior** [comment]
It is under the Simplified BSD License which should be compatible? ( Am no expert here)
It is designed. as a header-only library consisting of a single file that we could just drop into this repo.

**Comment by jbendes** [comment]
Yep that should be just fine! Not sure how this would fit in though. Considering the current blocking implementation requires the ability for the "popper" to go to sleep waiting on the "pusher". But that probably ties in closely with the asio conversation
